types/queue.go

```diff
@@ -0,0 +1,39 @@
+package types
+
+import "context"
+
+// Queue is the interface of a queue. Queue's methods are atomic
+// and goroutine-safe.
+type Queue interface {
+	// Enqueue adds a new item to the queue. It returns any error that
+	// was encountered in doing so, or if the context is cancelled.
+	Enqueue(ctx context.Context, value string) error
+
+	// Dequeue gets an Item from the queue. It returns the Item and any
+	// error encountered, or if the context is cancelled.
+	Dequeue(ctx context.Context) (*QueueItem, error)
+}
+
+// QueueItem contains a Value and embeds AckNack.
+type QueueItem struct {
+	// Value is the item's underlying value.
```

Comment:
Out of curiosity, what's the rationale for splitting the item and ack/nack functions? I tend to think of them as coupled, and I guess they are. I think reading the implementation as interfaces here is pretty straightforward and makes sense, but is a little confusing within the queue itself.

Reply:
I could have done:

```
type QueueItem interface {
    Ack(context.Context) error
    Nack(context.Context) error
    Value() string
}
```

Would that be preferable?

Reply:
Yeah, I think that would be a little clearer!